### PR TITLE
Add DualConnector

### PR DESCRIPTION
**Pull request by vbkaisetsu**

### CODE REVIEW

File: vibrato/src/dictionary/connector/dual_connector.rs [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        }
+
+        // Creates a MatrixConnector
+        let mut right_id_map = vec![0];
+        let mut left_id_map = vec![0];
+        let mut right_features_map = HashMap::new();
+        let mut left_features_map = HashMap::new();
+        right_features_map.insert(vec![U31::default(); col_size - SIMD_SIZE], 0);
+        left_features_map.insert(vec![U31::default(); col_size - SIMD_SIZE], 0);
+        for right_features in &right_ids_tmp {
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/dictionary/connector/dual_connector.rs]: fixed

File: vibrato/src/dictionary/connector/dual_connector.rs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+            if matrix_ids_set.contains(&i) {
+                matrix_ids.push(i);
+            } else {
+                raw_ids.push(i);
+            }
+        }
+
+        // Creates a MatrixConnector
+        let mut right_id_map = vec![0];
+        let mut left_id_map = vec![0];
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/dictionary/connector/dual_connector.rs]: I separated the function to `create_raw_connector` and `create_matrix_connector`.

File: vibrato/src/dictionary/connector/dual_connector.rs [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            col_size,
+            mut scorer_builder,
+        } = RawConnectorBuilder::from_readers(right_rdr, left_rdr, cost_rdr)?;
+        let scorer = scorer_builder.build();
+
+        // Split features into RawConnector and MatrixConnector
+        let matrix_ids_set =
+            Self::greedy_remover(SIMD_SIZE, &left_ids_tmp, &right_ids_tmp, col_size);
+        let mut raw_ids = vec![];
+        let mut matrix_ids = vec![];
```
[the remark this answers is not in the document]
Comment (vbkaisetsu) [reply vibrato/src/dictionary/connector/dual_connector.rs]: fixed

File: vibrato/src/dictionary/connector/dual_connector.rs [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    raw_right_ids: Vec<U31x8>,
+    raw_left_ids: Vec<U31x8>,
+    raw_scorer: Scorer,
+}
+
+impl DualConnector {
+    /// Removes feature templates so that the matrix size is smaller using greedy search
+    /// and returns a set of rest IDs.
+    pub fn remove_feature_templates_greedy(
+        raw_id_size: usize,
```
Comment (kampersanda) [inline vibrato/src/dictionary/connector/dual_connector.rs:29]:
```suggestion
        raw_idx_size: usize,
```
Could you use `idx` for consistency?

File: vibrato/src/dictionary/connector/dual_connector.rs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        left_ids: &[Vec<U31>],
+        right_ids: &[Vec<U31>],
+        col_size: usize,
+    ) -> HashSet<usize> {
+        let mut matrix_ids: HashSet<usize> = (0..col_size).collect();
```
Comment (kampersanda) [inline vibrato/src/dictionary/connector/dual_connector.rs:34]:
```suggestion
        let mut matrix_indices: HashSet<usize> = (0..col_size).collect();
```
ditto

File: vibrato/src/dictionary/connector/dual_connector.rs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+            let mut id_map = vec![0];
+            let mut features_map = HashMap::new();
+            features_map.insert(vec![U31::default(); col_size - SIMD_SIZE], 0);
+            for features in ids_tmp {
+                let mut feature_ids = vec![];
+                for &idx in matrix_indices {
+                    feature_ids.push(*features.get(idx).unwrap_or(&INVALID_FEATURE_ID));
+                }
+                let new_id = features_map.len();
+                let right_id = *features_map.entry(feature_ids).or_insert(new_id);
```
Comment (kampersanda) [inline vibrato/src/dictionary/connector/dual_connector.rs:88]: right_id -> orig_id?

File: vibrato/src/dictionary/connector/dual_connector.rs [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        let right_used_features: HashSet<_> = right_ids.iter().cloned().collect();
+        let left_used_features: HashSet<_> = left_ids.iter().cloned().collect();
+        for (i, left_map) in scorer_builder.trie.iter_mut().enumerate() {
+            if right_used_features.contains(&U31::new(u32::try_from(i).unwrap()).unwrap()) {
+                let left_indices: Vec<_> = left_map.keys().cloned().collect();
+                for id in left_indices {
+                    if !left_used_features.contains(&id) {
+                        left_map.remove(&id);
+                    }
+                }
```
Comment (kampersanda) [inline vibrato/src/dictionary/connector/dual_connector.rs:138]:
```suggestion
                for idx in left_indices {
                    if !left_used_features.contains(&idx) {
                        left_map.remove(&idx);
                    }
                }
```

File: vibrato/src/dictionary/connector/dual_connector.rs (hunk 4 above)
Comment (vbkaisetsu) [reply vibrato/src/dictionary/connector/dual_connector.rs]: I changed the name to `raw_feat_template_size`.

File: vibrato/src/dictionary/connector/dual_connector.rs (hunk 5 above)
Comment (vbkaisetsu) [reply vibrato/src/dictionary/connector/dual_connector.rs]: fixed

File: vibrato/src/dictionary/connector/dual_connector.rs (hunk 6 above)
Comment (vbkaisetsu) [reply vibrato/src/dictionary/connector/dual_connector.rs]: Fixed.
* `new_id` -> `new_conn_id`
* `right_id` -> `conn_id`

File: vibrato/src/dictionary/connector/dual_connector.rs (hunk 7 above)
Comment (vbkaisetsu) [reply vibrato/src/dictionary/connector/dual_connector.rs]: Renamed
* `left_indices` -> `left_feat_ids`
* `id` -> `feat_id`

**Review by kampersanda — APPROVED** [review]
LGTM, thanks!

**Pull request merged**